### ISSUE TITLE
[feature request] Allow embedding to use not only the last axis

### ISSUE THREAD

**Issue by arogozhnikov** [pytorch/pytorch#9025]
## Issue description

`torch.nn.Embedding` can provide one more argument - axis to use as an 'embedding dimensionality'

This will allow using `nn.Embedding` in sequential models.

## Code example

The following code is expected to work 
```
m = torch.nn.Sequential(
    torch.nn.Embedding(num_embeddings=26, embedding_dim=512, added_dim=1), 
    torch.nn.Conv1d(512, 512, kernel_size=3, padding=1),
)
```

New argument `added_dim=1` should be `-1` by default to agree with previous behavior.

**Comment by SsnL** [comment]
this changes output shape?

**Comment by arogozhnikov** [comment]
> this changes output shape?

Yes, it does. Below is detailed proposed behavior

```python
import torch
import numpy

def embed(x, embedding, added_dim=-1):
    x = embedding(x)
    d = numpy.arange(len(x.shape))
    d[added_dim:] = numpy.roll(d[added_dim:], 1)
    return x.permute(tuple(d))
```

And below is some demo how it affects shape:

```python
x = torch.zeros(1, 2, 3, 4).long()
for added_dim in range(-5, 5):
    print(added_dim, embed(x, torch.nn.Embedding(100, 30), added_dim=added_dim).shape)

output:
-5 torch.Size([30, 1, 2, 3, 4])
-4 torch.Size([1, 30, 2, 3, 4])
-3 torch.Size([1, 2, 30, 3, 4])
-2 torch.Size([1, 2, 3, 30, 4])
-1 torch.Size([1, 2, 3, 4, 30])
0 torch.Size([30, 1, 2, 3, 4])
1 torch.Size([1, 30, 2, 3, 4])
2 torch.Size([1, 2, 30, 3, 4])
3 torch.Size([1, 2, 3, 30, 4])
4 torch.Size([1, 2, 3, 4, 30])
```

**Comment by SsnL** [comment]
Can you use `permute` after embedding? I don't think there is enough use case to add this as an argument.

**Comment by arogozhnikov** [comment]
Ok, so this is how I am doing this now:

```python
from einops.layers.torch import Rearrange 

model = nn.Sequential(
    nn.Embedding(num_embeddings=26, embedding_dim=512),
    Rearrange('b t c -> b c t'),
    nn.Conv1d(512, 512, kernel_size=3, padding=1),
    <rest of the model>
)
```

All thanks to  [einops](https://github.com/arogozhnikov/einops).
I think we can close the issue now
